lib/accessmonitoring/monitor.go

```diff
@@ -0,0 +1,242 @@
+	// events.
+	Events types.Events
+}
+
+// CheckAndSetDefaults checks and sets default config values.
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Logger == nil {
+		c.Logger = slog.Default()
+	}
+	if c.Backend == nil {
+		return trace.BadParameter("backend: must be non-nil")
+	}
+	if c.Events == nil {
+		return trace.BadParameter("events: must be non-nil")
+	}
+	return nil
+}
+
+// AccessMonitor is an access monitoring service that monitors access events,
+// then executes the configured set of handlers for each event.
+type AccessMonitor struct {
+	cfg Config
+
+	// ruleHandlers contains the list of access monitoring rule event handlers.
+	ruleHandlers []EventHandler
+	// requestHandlers contains the list access request event handlers.
+	requestHandlers []EventHandler
+}
+
+// NewAccessMonitor returns a new access monitor.
+func NewAccessMonitor(cfg Config) (*AccessMonitor, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &AccessMonitor{
+		cfg: cfg,
+	}, nil
+}
+
+// AddAccessMonitoringRuleHandler appends the access monitoring rule event handler.
+// These handlers are executed whenever a new access monitoring rule event is
+// observed.
+func (s *AccessMonitor) AddAccessMonitoringRuleHandler(handler EventHandler) {
+	s.ruleHandlers = append(s.ruleHandlers, handler)
+}
+
+// AddAccessRequestHandler appends the access request handler.
+// These handlers are executed whenever a new access request event is observed.
+func (s *AccessMonitor) AddAccessRequestHandler(handler EventHandler) {
+	s.requestHandlers = append(s.requestHandlers, handler)
+}
+
+// Run runs the access monitor. Restarts on failure.
+func (s *AccessMonitor) Run(ctx context.Context) error {
+	const retryJitter = 10 * time.Second
+	const retryInterval = 30 * time.Second
+	const lockTTL = time.Minute
+
+	waitWithJitter := retryutils.SeventhJitter(retryJitter)
+	for {
+		err := backend.RunWhileLocked(ctx, backend.RunWhileLockedConfig{
+			LockConfiguration: backend.LockConfiguration{
+				Backend:            s.cfg.Backend,
+				LockNameComponents: []string{componentName},
+				TTL:                lockTTL,
+				RetryInterval:      retryInterval,
+			},
+		}, s.run)
+		if err != nil {
+			s.cfg.Logger.ErrorContext(
+				ctx,
+				"Encountered a fatal error, it will restart after backoff.",
+				"component", componentName,
+				"error", err,
+				"restart_after", waitWithJitter,
+			)
+		}
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(waitWithJitter):
+		}
+	}
+}
+
+func (s *AccessMonitor) run(ctx context.Context) error {
+	// Initialize the watcher.
+	const allowPartialSuccessFalse = false
+	watcher, err := s.cfg.Events.NewWatcher(ctx, types.Watch{
+		Name: componentName,
+		Kinds: []types.WatchKind{
+			{Kind: types.KindAccessRequest},
+			{Kind: types.KindAccessMonitoringRule},
+		},
+		AllowPartialSuccess: allowPartialSuccessFalse,
+	})
```

Comment:
This will cause the access monitoring service to fail if the plugin doesn't have the rights to read AMRs.
This might not be a bad thing, but will definitely be a breaking change and require long time users to update their roles. Are we comfortable with that?

If so, this should be part of the v18 release notes.

Reply:
I think it would be a good idea to make this change, as we continue to adopt access monitoring rules for more features. But, we don't need to make that decision now. It is probably a good idea to defer this change until a later time when we have more focus to ensure a smooth transition for users.